src/libraries/System.Net.Http/src/System/Threading/AsyncMutex.cs

```diff
@@ -0,0 +1,324 @@
+                            // The waiter was no longer in the list.  We must not cancel it.
+                            w = null;
+                        }
+                    }
+
+                    // If the waiter was in the list, we removed it under the lock and thus own
+                    // the ability to cancel it.  Do so.
+                    w?.Cancel();
+                }
+            }
+        }
+
+        /// <summary>Releases the mutex.</summary>
+        /// <remarks>The caller must logically own the mutex.  This is not validated.</remarks>
+        public void Exit()
+        {
+            if (Interlocked.Increment(ref _gate) < 1)
+            {
+                // This is the equivalent of:
+                //     _sem.Release();
+                // if _sem were to be constructed as `new SemaphoreSlim(0)`.
+                Contended();
+            }
+
+            void Contended()
+            {
+                Waiter? w;
+
+                lock (SyncObj)
+                {
+                    Debug.Assert(_lockedSemaphoreFull);
+
+                    // Wake up the next waiter in the list.
+                    w = _waitersHead;
+                    if (w != null)
+                    {
+                        // Remove the waiter.
+                        _waitersHead = w.Next;
+                        if (w.Next != null)
+                        {
+                            w.Next.Prev = null;
+                        }
+                        else
+                        {
+                            Debug.Assert(_waitersTail == w);
+                            _waitersTail = null;
+                        }
+                        w.Next = w.Prev = null;
+                    }
+                    else
+                    {
+                        // There wasn't a waiter.  Mark that the async lock is no longer full.
+                        Debug.Assert(_waitersTail is null);
+                        _lockedSemaphoreFull = false;
+                    }
+                }
+
+                // Either there wasn't a waiter, or we got one and successfully removed it from the list,
+                // at which point we own the ability to complete it.  Do so.
+                w?.Set();
+            }
+        }
+
+        /// <summary>Creates a canceled ValueTask.</summary>
+        /// <remarks>Separated out to reduce asm for this rare path in the call site.</remarks>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static ValueTask FromCanceled(CancellationToken cancellationToken) =>
+            new ValueTask(Task.FromCanceled(cancellationToken));
+
+        /// <summary>Represents a waiter for the mutex.</summary>
+        /// <remarks>Implemented as a reusable backing source for a value task.</remarks>
+        private sealed class Waiter : IValueTaskSource
+        {
+            private ManualResetValueTaskSourceCore<bool> _mrvtsc; // mutable struct; do not make this readonly
```

Comment:
Aside from the PR:
I wonder if there's an attribute / analyzer-combo to put emphasis on the mutable struct, so the comment isn't necessary. Or is this too much an edge-case so it's not worth having something like this?

Reply:
https://github.com/dotnet/roslyn-analyzers/issues/2811

Reply:
Thanks.